From aa4e3ce99f4331c0d0e4a9a9147b7e6eb1c787e2 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Fri, 20 Sep 2024 05:58:30 +0300
Subject: [PATCH] [FSW-9512] fs1: set validation pass

This patch set validation pass flag just before
bootrom handover control to ppboot.
Ppboot transition is the place where we see at
least one manifest is enumerated, validated
properly and atleast one ppboot image is
prefetched, validated properly.
Store same manifest index to persistent dccm
and cold flops which is going to load ppboot.
Recovery is the flag which hold opposite value
to validation pass, to route recovery when
validation fail.

Change-Id: I37e5141fb215356d2e27604fb6f0ade4d1f9a5d5
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 17f4a2f..51d19e7 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -392,3 +392,27 @@
 	if (boot_rom_global.config2_fuse & EFUSE_SYSCFG2_RMA_MODE)
 		boot_rom_global.sec_data.rma_enabled = 1;
 }
+
+void boot_rom_validation_status(bool validation_pass)
+{
+	struct fs1_cold_rst_data cold_rst_data;
+
+	/* validation pass, set validation flag and clear recovery or viceversa */
+	cold_rst_data.data = regmap_readl(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_COLD_RST_REG_2));
+	cold_rst_data.recovery_flag = (validation_pass ? 0 : 1);
+	cold_rst_data.validation_flag = (validation_pass ? 1 : 0);
+	regmap_writel(cold_rst_data.data, IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_COLD_RST_REG_2));
+}
+
+void boot_rom_set_manifest_index(enum hl_fw_manifest_type man_idx)
+{
+	struct fs1_cold_rst_data cold_rst_data;
+
+	boot_rom_global.valid_manifest_idx = man_idx;
+	cold_rst_data.data = regmap_readl(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_COLD_RST_REG_2));
+	if (cold_rst_data.valid_manifest_idx == man_idx)
+		return;
+
+	cold_rst_data.valid_manifest_idx = man_idx;
+	regmap_writel(cold_rst_data.data, IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_COLD_RST_REG_2));
+}
diff --git a/common/boot_flow.c b/common/boot_flow.c
index c272687..1e0440a 100644
--- a/common/boot_flow.c
+++ b/common/boot_flow.c
@@ -144,15 +144,15 @@
 
 	/* TODO: MemCopy to PSOC FW SRAM if required */
 
-	/* TODO: Set validation pass and set manifest idx in global */
+	boot_rom_validation_status(true);
+
+	boot_rom_set_manifest_index(man_idx);
 
 	/* TODO: Set other manifest as invalid */
 
 	boot_rom_set_boot_info();
 
 	/* TODO: TPM PCR extend */
-
-	/* TODO: Clear recovery status */
 
 	if (desc.start_addr) {
 		/* TODO: Start CPLD watchdog timer */
diff --git a/common/common.h b/common/common.h
index b40f7d7..439872b 100644
--- a/common/common.h
+++ b/common/common.h
@@ -33,5 +33,7 @@
 bool boot_rom_in_multi_die_mode(void);
 uint32_t boot_rom_get_arc_addr(uint64_t addr);
 uint32_t boot_rom_verify_manifest(enum hl_fw_manifest_type manifest_idx);
+void boot_rom_validation_status(bool validation_pass);
+void boot_rom_set_manifest_index(enum hl_fw_manifest_type manifest_idx);
 
 #endif /* _COMMON_H_ */
